modules/intensity_transform/src/BIMEF.cpp

```diff
@@ -0,0 +1,560 @@
+    diag_idx_zero << 0;
+    Eigen::SparseMatrix<float> A = (Ax + Ay) + Eigen::SparseMatrix<float>((Ax + Ay).transpose()) + spdiags(D, diag_idx_zero, k, k);
+
+    //CG solver of Eigen
+    Eigen::ConjugateGradient<Eigen::SparseMatrix<float>, Eigen::Lower|Eigen::Upper, Eigen::IncompleteCholesky<float> > cg;
+    cg.setTolerance(0.1f);
+    cg.setMaxIterations(50);
+    cg.compute(A);
+    Mat_<float> img_t = img.t();
+    Eigen::Map<const Eigen::VectorXf> tin(img_t.ptr<float>(), img_t.rows*img_t.cols);
+    Eigen::VectorXf x = cg.solve(tin);
+
+    Mat_<float> tout(img.rows, img.cols);
+    for (int j = 0; j < img.cols; j++)
+    {
+        for (int i = 0; i < img.rows; i++)
+        {
+            tout(i,j) = x(j*img.rows + i);
+        }
+    }
+
+    return tout;
+}
+
+Mat_<float> tsmooth(const Mat_<float>& src, float lambda=0.01, float sigma=3.0, float sharpness=0.001)
+{
+    Mat_<float> W_h, W_v;
+    computeTextureWeights(src, sigma, sharpness, W_h, W_v);
+
+    Mat_<float> S = solveLinearEquation(src, W_h, W_v, lambda);
+
+    return S;
+}
+
+Mat_<float> rgb2gm(const Mat_<Vec3f>& I)
+{
+    Mat_<float> gm(I.rows, I.cols);
+    for (int i = 0; i < I.rows; i++)
+    {
+        for (int j = 0; j < I.cols; j++)
+        {
+            gm(i,j) = std::pow(I(i,j)[0]*I(i,j)[1]*I(i,j)[2], 1/3.0);
+        }
+    }
+
+    return gm;
+}
+
+Mat_<float> applyK(const Mat_<float>& I, float k, float a=-0.3293f, float b=1.1258f) {
+    float beta = std::exp((1 - std::pow(k, a)) * b);
+    float gamma = std::pow(k, a);
+
+    Mat_<float> J(I.size());
+    pow(I, gamma, J);
+    J = J*beta;
+
+    return J;
+}
+
+Mat_<Vec3f> applyK(const Mat_<Vec3f>& I, float k, float a=-0.3293f, float b=1.1258f, float offset=0) {
+    float beta = std::exp((1 - std::pow(k, a)) * b);
+    float gamma = std::pow(k, a);
+
+    Mat_<Vec3f> J(I.size());
+    pow(I, gamma, J);
+    for (int i = 0; i < I.rows; i++)
+    {
+        for (int j = 0; j < I.cols; j++)
+        {
+            J(i,j)[0] = J(i,j)[0] * beta + offset;
+            J(i,j)[1] = J(i,j)[1] * beta + offset;
+            J(i,j)[2] = J(i,j)[2] * beta + offset;
+        }
+    }
+
+    return J;
+}
+
+float entropy(const Mat_<float>& I)
+{
+    Mat_<uchar> I_uchar;
+    I.convertTo(I_uchar, CV_8U, 255);
+
+    std::vector<Mat> planes;
+    planes.push_back(I_uchar);
+    Mat_<float> hist;
+    const int histSize = 256;
+    float range[] = { 0, 256 };
+    const float* histRange = { range };
+    calcHist(&I_uchar, 1, NULL, Mat(), hist, 1, &histSize, &histRange);
+
+    Mat_<float> hist_norm = hist / cv::sum(hist)[0];
+
+    float E = 0;
+    for (int i = 0; i < hist_norm.rows; i++)
+    {
+        if (hist_norm(i,0) > 0)
+        {
+            E += hist_norm(i,0) * std::log2(hist_norm(i,0));
+        }
+    }
+
+    return -E;
+}
+
+template <typename T> int sgn(T val)
+{
+    return (T(0) < val) - (val < T(0));
+}
+
+double minimize_scalar_bounded(const Mat_<float>& I, double begin, double end,
```

Comment:
Code ported from [here](https://github.com/scipy/scipy/blob/v1.4.1/scipy/optimize/optimize.py#L1753-L1894).

Reply:
Thank you so much for your effort!